### PR TITLE
Reimplement ConstProp using DataflowConstProp

### PR DESCRIPTION
**Pull request by cjgillot**
This PR attempts to remove the current ConstProp optimization, and replace it by an implementation based on DataflowConstProp.

The current ConstProp twists the MIR interpreter to do things it is not designed to do, including manipulating generic types. Meanwhile, @jachris has implemented a new DataflowConstProp pass that only uses a few chosen methods for arithmetic.

The main change concerns how propagatable places and operations are tracked. In the current ConstProp, all places are tracked and all operations executed, and we removed wrong or non-propagatable results. This new implementation does the opposite : place tracking is opt-in, and we only propagate supported operations. I believe this approach to be less risky from a soundness point of view.

As a consequence, this new implementation is more restricted. The main limitation is that we stop propagating inside a basic block, and only propagate SSA locals. This makes the implementation simpler, and is more consistent with this "opt-in" logic.

Some operations are added inside this PR (transmutes, algebraic simplifications, `Len` operations). The handling of statics and non-scalar consts are left to a follow-up PR.

This PR *does not* remove the const-prop lints (arithmetic overflow and unconditional panic), as I need to design a way to pass more span information around.

There has been quite a few perf runs below. Marking as ready for review, since I removed the massive regressions on diesel/keccak/serde. Kept as waiting-on-author until I bless CI.

Regressed optimizations:
- [ ] array indexing;
- [ ] non-scalar constants;
- [ ] assignment of immutable statics;
- [ ] creating rvalues with non-scalar types;
- [ ] integer-to-pointer transmutes.

Based on https://github.com/rust-lang/rust/pull/110728, https://github.com/rust-lang/rust/pull/110732, https://github.com/rust-lang/rust/pull/110820

### CODE REVIEW

**Comment by cjgillot** [comment]
r? wg-mir-opt

**Comment by cjgillot** [comment]
r? compiler

File: tests/mir-opt/const_prop/array_index.main.ConstProp.32bit.diff [hunk 1]
```diff
@@ -23,12 +23,12 @@
...
 +         _4 = const 4_usize;              // scope 0 at $DIR/array_index.rs:+1:18: +1:33
 +         _5 = const true;                 // scope 0 at $DIR/array_index.rs:+1:18: +1:33
-+         assert(const true, "index out of bounds: the length is {} but the index is {}", move _4, _3) -> bb1; // scope 0 at $DIR/array_index.rs:+1:18: +1:33
++         assert(const true, "index out of bounds: the length is {} but the index is {}", const 4_usize, const 2_usize) -> bb1; // scope 0 at $DIR/array_index.rs:+1:18: +1:33
       }
   
       bb1: {
 -         _1 = _2[_3];                     // scope 0 at $DIR/array_index.rs:+1:18: +1:33
-+         _1 = const 2_u32;                // scope 0 at $DIR/array_index.rs:+1:18: +1:33
++         _1 = _2[2 of 3];                 // scope 0 at $DIR/array_index.rs:+1:18: +1:33
```
Comment (oli-obk) [inline tests/mir-opt/const_prop/array_index.main.ConstProp.32bit.diff:31]: looks like dataflow const prop can't handle indexing yet?

File: tests/mir-opt/const_prop/mutable_variable.main.ConstProp.diff [hunk 2]
```diff
@@ -17,8 +17,7 @@
           _1 = const 42_i32;               // scope 0 at $DIR/mutable_variable.rs:+1:17: +1:19
           _1 = const 99_i32;               // scope 1 at $DIR/mutable_variable.rs:+2:5: +2:11
           StorageLive(_2);                 // scope 1 at $DIR/mutable_variable.rs:+3:9: +3:10
--         _2 = _1;                         // scope 1 at $DIR/mutable_variable.rs:+3:13: +3:14
-+         _2 = const 99_i32;               // scope 1 at $DIR/mutable_variable.rs:+3:13: +3:14
+          _2 = _1;                         // scope 1 at $DIR/mutable_variable.rs:+3:13: +3:14
```
Comment (oli-obk) [inline tests/mir-opt/const_prop/mutable_variable.main.ConstProp.diff:20]: what happened here?

File: tests/mir-opt/const_prop/read_immutable_static.main.ConstProp.diff [hunk 3]
```diff
@@ -20,18 +20,15 @@
...
           _5 = const {alloc1: &u8};        // scope 0 at $DIR/read_immutable_static.rs:+1:19: +1:22
                                            // mir::Constant
                                            // + span: $DIR/read_immutable_static.rs:7:19: 7:22
                                            // + literal: Const { ty: &u8, val: Value(Scalar(alloc1)) }
--         _4 = (*_5);                      // scope 0 at $DIR/read_immutable_static.rs:+1:19: +1:22
--         _1 = Add(move _2, move _4);      // scope 0 at $DIR/read_immutable_static.rs:+1:13: +1:22
-+         _4 = const 2_u8;                 // scope 0 at $DIR/read_immutable_static.rs:+1:19: +1:22
-+         _1 = const 4_u8;                 // scope 0 at $DIR/read_immutable_static.rs:+1:13: +1:22
+          _4 = (*_5);                      // scope 0 at $DIR/read_immutable_static.rs:+1:19: +1:22
+          _1 = Add(move _2, move _4);      // scope 0 at $DIR/read_immutable_static.rs:+1:13: +1:22
```
Comment (oli-obk) [inline tests/mir-opt/const_prop/read_immutable_static.main.ConstProp.diff:31]: another const prop regression

File: tests/mir-opt/const_prop/return_place.add.ConstProp.diff [hunk 4]
```diff
@@ -6,9 +6,8 @@
       let mut _1: (u32, bool);             // in scope 0 at $DIR/return_place.rs:+1:5: +1:10
   
       bb0: {
--         _1 = CheckedAdd(const 2_u32, const 2_u32); // scope 0 at $DIR/return_place.rs:+1:5: +1:10
+          _1 = CheckedAdd(const 2_u32, const 2_u32); // scope 0 at $DIR/return_place.rs:+1:5: +1:10
 -         assert(!move (_1.1: bool), "attempt to compute `{} + {}`, which would overflow", const 2_u32, const 2_u32) -> bb1; // scope 0 at $DIR/return_place.rs:+1:5: +1:10
-+         _1 = const (4_u32, false);       // scope 0 at $DIR/return_place.rs:+1:5: +1:10
 +         assert(!const false, "attempt to compute `{} + {}`, which would overflow", const 2_u32, const 2_u32) -> bb1; // scope 0 at $DIR/return_place.rs:+1:5: +1:10
```
Comment (oli-obk) [inline tests/mir-opt/const_prop/return_place.add.ConstProp.diff:11]: I'm guessing checked math doesn't work yet?

File: tests/mir-opt/const_prop/transmute.unreachable_ref.ConstProp.32bit.diff [hunk 5]
```diff
@@ -15,11 +15,7 @@
       bb0: {
           StorageLive(_1);                 // scope 0 at $DIR/transmute.rs:+0:38: +3:2
           StorageLive(_2);                 // scope 0 at $DIR/transmute.rs:+1:9: +1:10
--         _2 = const 1_usize as &Never (Transmute); // scope 2 at $DIR/transmute.rs:+1:30: +1:48
-+         _2 = const {0x1 as &Never};      // scope 2 at $DIR/transmute.rs:+1:30: +1:48
-+                                          // mir::Constant
-+                                          // + span: no-location
-+                                          // + literal: Const { ty: &Never, val: Value(Scalar(0x00000001)) }
+          _2 = const 1_usize as &Never (Transmute); // scope 2 at $DIR/transmute.rs:+1:30: +1:48
```
Comment (oli-obk) [inline tests/mir-opt/const_prop/transmute.unreachable_ref.ConstProp.32bit.diff:18]: transmute casts also don't seem to work yet

File: tests/mir-opt/const_prop/mutable_variable.main.ConstProp.diff (hunk 2 above)
Comment (cjgillot) [reply tests/mir-opt/const_prop/mutable_variable.main.ConstProp.diff:20]: `_1` is assigned twice, and this pass is the cheap version limited to SSA locals.

File: tests/mir-opt/const_prop/read_immutable_static.main.ConstProp.diff (hunk 3 above)
Comment (cjgillot) [reply tests/mir-opt/const_prop/read_immutable_static.main.ConstProp.diff:31]: We don't try to evaluate through non-scalar consts and statics. This is one of the next steps to implement.

File: tests/mir-opt/const_prop/return_place.add.ConstProp.diff (hunk 4 above)
Comment (cjgillot) [reply tests/mir-opt/const_prop/return_place.add.ConstProp.diff:11]: It does work. All uses of `_1.0` and `_1.1` are propagated. However, we don't replace the `CheckedAdd` rvalue with the evaluated tuple.

**Comment by cjgillot** [comment]
I added a list of regressions to the PR description.

**Comment by apiraino** [comment]
@cjgillot can you have a look into a rebase?  thanks!

@rustbot author

**Comment by cjgillot** [comment]
I've put this PR on hold, because I'm not quite sure of the direction I'm going with it. The DataflowConstProp is tailored to propagating scalars, and that makes it more difficult to propagate non scalar information.
I'm leaving this PR as waiting on author until I get to agree with myself.

File: tests/coverage-map/status-quo/issue-84561.cov-map [hunk 6]
```diff
@@ -1,11 +1,10 @@
 Function name: <issue_84561::Foo as core::cmp::PartialEq>::eq
-Raw bytes (14): 0x[01, 01, 00, 02, 01, 04, 0a, 00, 13, 00, 00, 0a, 00, 13]
+Raw bytes (9): 0x[01, 01, 00, 01, 01, 04, 0a, 00, 13]
```
Comment (cjgillot) [inline tests/coverage-map/status-quo/issue-84561.cov-map:2]: @Zalathar This commit blesses the coverage map. However, I have no idea how to check that it is still correct. What do you recommend?

Comment (Zalathar) [reply tests/coverage-map/status-quo/issue-84561.cov-map:2]: As a rule of thumb, any PR that doesn't touch coverage-specific code should feel free to just re-bless these.

(They're unfortunately quite sensitive to changes in MIR lowering/optimization, in a way that tends to result in different but equivalent mappings.)

Comment (Zalathar) [reply tests/coverage-map/status-quo/issue-84561.cov-map:2]: In particular, as long as the `run-coverage` tests still pass, then any changes to the mappings are probably just noise.

(Those tests only run if the profiler runtime is enabled in `config.toml`, so they don't run by default in dev builds or on PRs, but they do run in the full CI before merging.)

Comment (Zalathar) [reply tests/coverage-map/status-quo/issue-84561.cov-map:2]: Looking at the actual diff, it seems like some MIR statements that were previously being optimized out are no longer being optimized out. Nothing to worry about from a coverage perspective.

**Comment by cjgillot** [comment]
I'd like to close this PR and pursue another direction for `ConstProp`.

https://github.com/rust-lang/rust/pull/109597 creates a general framework for value analyses. I believe that implementing constant-folding in this pass can achieve the same level as what this PR proposes, while being more flexible:
- we'd only use the interpreter arithmetic API, like dataflow const-prop does, and allow removing `ConstPropNonsense`;
- that pass would only propagate whole values, simplifying state storage, and leaving the per-place to the dataflow version;
- that pass would leverage arithmetic identities over unknown values, which const-prop cannot do;
- that pass may be able to avoid duplicating constant allocations.

**Comment by RalfJung** [comment]
> that pass would leverage arithmetic identities over unknown values, which const-prop cannot do;

const-prop can simplify some things here, right, like x*0 -> 0 or x+0 -> x?

**Comment by cjgillot** [comment]
Yes, but it cannot do `x - x = 0`, which #116012 could (in https://github.com/rust-lang/rust/pull/111344).

**Pull request closed without merging**